### DISCUSSION TITLE
Refresh document diagnostics

### DISCUSSION THREAD

**Discussion by luiz290788** [eclipse-langium/langium#987] [Q&A] [answered]
Is there a way to refresh the diagnostics of a document?

I have document that references other documents and I'm loading these other documents after the first one is parsed. But then the diagnostics of the original one just takes the first document into consideration. When I make any edit in the original document then diagnostics is fine. I think that there is some sort of race condition happening. If I could refresh the diagnostics for the first document, then I could force it right after adding all other documents. One interesting thing is that Langium is not firing `DocumentState.Validated` for the second document, but if I try to get the document with `services.workspace.LangiumDocuments.getOrCreateDocument(uri)` the document is there.

I'm being very abstract here, I will work on an example to of this later today isolating the issue.

**Comment by msujew** [comment, marked as answer]
I imagine your outside of the language server usecase? Note that lifecycle handling is a bit more manual once you move away from the language server. The [langium-cli](https://github.com/langium/langium/blob/479fedbb7a8dd5ecf111fe3cf5a2979731b14ce2/packages/langium-cli/src/generate.ts#L159-L178) has a functionality which only updates document validations once all referenced documents have been loaded by the CLI. Is this similar to what you're trying to accomplish?

Anyway, you can always refresh them manually by just running the document through the `DocumentValidator`.

**Reply by luiz290788** [reply]
I'm in the language server. I don't know why it is does not update automatically. I was able to make it work with `services.workspace.DocumentBuilder.update([uri], [])`. I don't know if that is the best way to do it.

Interesting use case of the cli. It is very similar to what I'm doing, but inside of the browser using a language server. I get the AST of a document and start loading the other documents referenced by it. After that I want to force an update in the main document.

I will try to isolate the issue later to show you, maybe it is a bug.

**Reply by msujew** [reply]
I see, I mainly meant the node based language server. Are you using an override of the `FileSystemProvider` for the browser use case? Using that should initialize the language server correctly. Of course, passing files manually works as well, though you really need to take care of race conditions and so on. Please make sure to use the `Mutex` in the shared workspace services.

**Reply by luiz290788** [reply]
I used this for the file system:

```ts
const { shared } = createMyLanguageServices({ connection, ...EmptyFileSystem })
```

Do I need to do something else?

BTW, what is the correct schema that I need to use for memory modes? `inmemory` or `memory`?

**Reply by msujew** [reply]
> BTW, what is the correct schema that I need to use for memory modes? inmemory or memory?

Doesn't matter, there's no URI schema which has any special meaning in Langium.

> Do I need to do something else?

Not necessarily, but if you have something like an online-workspace, it might make sense to create your own implementation of a `FileSystemProvider` to provide this workspace to the `WorkspaceManager`. This is the well-tested solution that ensures workspace initialization happens correctly.
